work-pool/src/WorkPool.hs

```diff
@@ -0,0 +1,78 @@
+module WorkPool
+  ( Config(..)
+  , Pool
+  , pushJob
+  , runPool
+  )
+where
+
+import MPrelude
+import Prelude (succ)
+
+import qualified UnliftIO.Async as UnliftIO
+import qualified UnliftIO.STM   as UnliftIO
+
+-- | Worker pool configuration
+data Config a = Config
+  { produceJobs :: forall m . MonadUnliftIO m => Pool a -> m ()
+  -- ^ function called from the main thread producing work, use `pushJob` to
+  -- create workable jobs.
+  , queueSize   :: Natural
+  -- ^ maximum size of the jobs queued
+  , workerCount :: Natural
+  -- ^ number of workers to boot
+  , workerRun   :: forall m . MonadUnliftIO m => Natural -> m (a -> m ())
+  -- ^ function called when a worker is booted, argument is the worker index,
+  -- returns an action to be called per job assigned to this worker.
+  }
+
+-- Internal queue event, supplying job or quitting the worker
+data Event a = Quit | Job a
+
+-- | Running pool
+newtype Pool a = Pool
+  { queue :: UnliftIO.TBQueue (Event a)
+  }
+
+-- | Add (dynamically) created a job to the pool
+--
+-- This function will block if the max queue size would be overflown.
+-- As the workers create space in the queue this function will unblock.
+pushJob :: MonadIO m => Pool a -> a -> m ()
+pushJob Pool{..} item
+  = UnliftIO.atomically
+  $ UnliftIO.writeTBQueue queue (Job item)
+
+-- | Run worker pool with specified config
+--
+-- The function will return if either:
+-- * the `produceJobs` function returns
+-- * a worker or throws an error
+-- * the producer throws an error.
+--
+-- Care is taken to not leak threads via the use if `withAsync` from the `async` package.
+runPool :: forall a m . MonadUnliftIO m => Config a -> m ()
+runPool Config{..} = do
+  pool@Pool{..} <- UnliftIO.atomically $ do
+    queue <- UnliftIO.newTBQueue queueSize
+    pure Pool{..}
+
+  boot pool $ \handlers -> do
+    produceJobs pool                                       -- supply jobs
+    UnliftIO.atomically $ UnliftIO.writeTBQueue queue Quit -- signal workers to gracefully exit
+    traverse_ UnliftIO.wait handlers                       -- wait for workers to gracefully exit
+  where
+    boot Pool{..} withAsyncHandlers = go 0 []
```

Comment:
nit-pick: I think a type signature for the `boot `function would aid readability. I spent a few cycles putting it together in my head.